File: app/src/main/java/com/uncmorfi/balance/dialogs/UserOptionsDialog.java
package com.uncmorfi.balance.dialogs;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatDialogFragment;

import com.uncmorfi.R;
import com.uncmorfi.balance.backend.BalanceBackend;
import com.uncmorfi.balance.model.User;

/**
 * Muestra las opciones disponibles para efectuar sobre un usuario.
 */
public class UserOptionsDialog extends AppCompatDialogFragment {
    public static final String ARG_USER = "user";

    /**
     * @param user Puede no contener todos los datos del usuario,
     *             pero necesita además de los datos que necesitan las demás opciones:
     *             {@link User#getCard()}
     *             {@link User#getPosition()}
     */
    public static UserOptionsDialog newInstance(User user) {
        Bundle args = new Bundle();

        args.putSerializable(ARG_USER, user);

        UserOptionsDialog fragment = new UserOptionsDialog();
        fragment.setArguments(args);

        return fragment;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final AlertDialog.Builder builder = new AlertDialog.Builder(getContext());

        final CharSequence[] items = new CharSequence[5];
        items[0] = getString(R.string.balance_user_options_update);
        items[1] = getString(R.string.balance_user_options_delete);
        items[2] = getString(R.string.balance_user_options_copy);
        items[3] = getString(R.string.balance_user_options_barcode);
        items[4] = getString(R.string.balance_user_options_set_name);

        final User user = (User) getArguments().getSerializable(ARG_USER);
        final BalanceBackend backend = BalanceBackend.getInstance(getContext());

        if (backend != null && user != null) {
            builder.setTitle(getString(R.string.balance_user_options_title))
                    .setItems(items, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            switch (which) {
                                case 0:
                                    backend.updateBalanceOfUser(
                                            user.getCard(), new int[]{user.getPosition()});
                                    break;
                                case 1:
                                    DeleteUserDialog.newInstance(user)
                                            .show(getFragmentManager(), "DeleteUserDialog");
                                    break;
                                case 2:
                                    backend.copyCardToClipboard(user.getCard());
                                    break;
                                case 3:
                                    BarcodeDialog.newInstance(user)
                                            .show(getFragmentManager(), "BarcodeDialog");
                                    break;
                                case 4:
                                    SetNameDialog.newInstance(user)
                                            .show(getFragmentManager(), "SetNameDialog");
                                    break;
                                default:
                                    break;
                            }
                        }
                    });
        }
        return builder.create();
    }

    @Override
    public void onPause() {
        super.onPause();
        dismiss();
    }
}

File: app/src/main/java/com/uncmorfi/balance/dialogs/UserOptionsDialog.kt
package com.uncmorfi.balance.dialogs

import android.app.Dialog
import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.uncmorfi.R
import com.uncmorfi.data.persistence.entities.User
import com.uncmorfi.reservations.CaptchaDialog
import com.uncmorfi.reservations.ReserveOptionsDialog
import com.uncmorfi.shared.*
import kotlinx.android.synthetic.main.dialog_user.view.*
import kotlinx.android.synthetic.main.dialog_user.view.userBalance
import kotlinx.android.synthetic.main.dialog_user.view.userName

/**
 * Muestra las opciones disponibles para efectuar sobre un usuario.
 */
class UserOptionsDialog : BaseDialogHelper() {


    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        super.init()

        val v = View.inflate(context, R.layout.dialog_user, null)

        v.userName.text = user.name
        v.userEmail.text = user.email
        v.userType.text = user.type
        v.userBalance.text = user.balance.toMoneyFormat()
        v.userPrice.text =
            getString(R.string.balance_user_options_price, user.price.toMoneyFormat())

        Glide.with(requireContext())
            .load(user.image)
            .placeholder(R.drawable.ic_account)
            .diskCacheStrategy(DiskCacheStrategy.ALL)
            .apply(RequestOptions.circleCropTransform())
            .into(v.userPhoto)

        builder.setView(v)

        return builder.create()
    }

    private var reserveCached = false

    fun onCreateDialog2(savedInstanceState: Bundle?): Dialog {
        reservationInit()
        val items = arrayOf(
            getString(R.string.balance_user_options_update),
            getString(R.string.balance_user_options_reserve),
            getString(R.string.balance_user_options_delete),
            getString(R.string.balance_user_options_copy),
//                getString(R.string.balance_user_options_barcode),
            getString(R.string.balance_user_options_set_name)
        )

        builder.setTitle(getString(R.string.balance_user_options_title))
            .setItems(items) { _, which ->
                when (which) {
                    0 -> sendResult(which, user)
                    1 -> reservation()
                    2 -> DeleteUserDialog
                        .newInstance(this, 0, user)
                        .show(parentFragmentManager, "DeleteUserDialog")
                    3 -> {
                        context?.copyToClipboard("card", user.card)
                        viewModel.status.value = StatusCode.COPIED
                    }
//                        4 -> startActivity(BarcodeActivity.intent(context!!, user))
                    4 -> SetNameDialog
                        .newInstance(this, 0, user)
                        .show(parentFragmentManager, "SetNameDialog")
                }
            }
        return builder.create()
    }

    private fun reservationInit() {
        viewModel.reservation.observe(this, Observer {
            reserveCached = reserveCached || (it == ReserveStatus.CACHED)
        })
        viewModel.reserveIsCached(user)
    }

    private fun reservation() {
        if (reserveCached) {
            ReserveOptionsDialog
                .newInstance(this, 0, user)
                .show(parentFragmentManager, "ReserveOptionsDialog")
        } else {
            CaptchaDialog
                .newInstance(this, 0, user)
                .show(parentFragmentManager, "CaptchaDialog")
        }
    }

    companion object {
        fun newInstance(fragment: Fragment, code: Int, user: User): UserOptionsDialog {
            return newInstance(::UserOptionsDialog, fragment, code, user)
        }
    }

}
